### PR TITLE
refactor(okta): model AWS role mappings

### PR DESCRIPTION
**Pull request by kunaals**
### Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [x] Refactoring (no functional changes)
- [x] Documentation update

### Summary

Model Okta group-to-AWS role mappings as scoped MatchLinks instead of handwritten Cypher. This makes the canonical `(:OktaGroup)-[:HAS_ROLE]->(:AWSRole)` relationship available to schema introspection and generated Okta and AWS documentation.

The sync continues to create the deprecated `ALLOWED_BY` relationship for backward compatibility. It will be removed in Cartography v1. The MatchLinks are scoped to their Okta organization, stale relationships are cleaned up, and a migration cleanup removes unscoped relationships created by earlier versions.

### Related issues or links

- Follow-up to https://github.com/cartography-cncf/cartography/pull/1243#issuecomment-5351995038

### How was this tested?

- Synced Okta from a live tenant in read-only provider mode into an isolated local Neo4j database.
- Used a sanitized, database-only mapping to exercise creation, stale cleanup, and recreation against the live-synced Okta graph and local AWS Identity Center nodes. The tenant's current group names did not match the configured account-and-role naming pattern, so the validation did not rely on naturally occurring cross-module edges.

Sanitized validation output:

```text
INFO:cartography.intel.okta.awssaml:Syncing Okta SAML Integration
created=ALLOWED_BY:1,HAS_ROLE:1
stale_after_unmap=0
recreated=ALLOWED_BY:1,HAS_ROLE:1
invalid_scope=0
```

#### General

- [x] I have read the [contributing guidelines](https://github.com/cartography-cncf/cartography/blob/master/CONTRIBUTING.md).
- [x] The linter passes locally (`make test_lint`).
- [x] I have added/updated tests that prove my fix is effective or my feature works.
- [x] Every commit includes a DCO `Signed-off-by` trailer.

#### Proof of functionality

- [ ] Screenshot showing the graph before and after changes.
- [x] New or updated unit/integration tests.

#### If you are adding or modifying an intel connector

- [x] Tested the connector against a real cloud, SaaS, or provider environment.
- [x] Included sanitized Cartography sync logs demonstrating successful synchronization of the new/modified connector behavior.
- [x] Removed credentials, personal data, account identifiers, and other sensitive values from the evidence.

#### If you are changing a node or relationship

- [x] Updated model docstrings and `PropertyRef.description` values used to generate schema documentation.
- [x] Built the documentation with `uv run ./docs/build.sh`.

#### If you are implementing a new intel module

- [ ] Used the NodeSchema [data model](https://docs.cartography.dev/dev/writing-intel-modules.html#defining-a-node).

### Notes for reviewers

The deprecated `ALLOWED_BY` compatibility schema is intentionally excluded from generated documentation. The canonical `HAS_ROLE` relationship appears in both the Okta and AWS generated schema pages.

### CODE REVIEW

File: cartography/data/jobs/cleanup/okta_import_cleanup.json [hunk 1]
```diff
@@ -30,6 +30,12 @@
       "iterationsize": 100,
       "__comment__": "Migration: drop deprecated Okta-to-Kubernetes group mapping edges, replaced by MEMBER_OF"
     },
+    {
+      "query": "MATCH (:OktaOrganization{id: $OKTA_ORG_ID})-[:RESOURCE]->(:OktaGroup)-[r:ALLOWED_BY]->(:AWSRole) WHERE r._sub_resource_label IS NULL WITH r LIMIT $LIMIT_SIZE DELETE r",
```
Comment (jychp) [inline cartography/data/jobs/cleanup/okta_import_cleanup.json]: drop the `$OKTA_ORG_ID` scoping from this migration statement. The legacy loader matched `(app:OktaApplication{name:'amazon_aws'})--(group:OktaGroup)` with no org filter, so unscoped `ALLOWED_BY` edges exist on groups whose OktaOrganization is no longer synced, and this statement can never reach them.

File: cartography/models/okta/awssaml.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    direction: LinkDirection = LinkDirection.OUTWARD
+    rel_label: str = "HAS_ROLE"
+    properties: OktaGroupToAWSRoleRelProperties = OktaGroupToAWSRoleRelProperties()
+
+
+@dataclass(frozen=True)
+# DEPRECATED: replaced by the canonical (:UserGroup)-[:HAS_ROLE]->(:PermissionRole)
+# edge (OktaGroupToAWSRoleHasRoleMatchLink). Kept for backward compatibility and
+# will be removed in v1.0.0.
+class OktaGroupToAWSRoleAllowedByMatchLink(CartographyRelSchema):
```
Comment (jychp) [inline cartography/models/okta/awssaml.py]: Subclass `OktaGroupToAWSRoleHasRoleMatchLink` and override only `rel_label`. The two schemas duplicate every matcher, sub-resource and properties field today and will drift.

File: cartography/intel/okta/awssaml.py [hunk 3]
```diff
@@ -97,32 +108,41 @@ def query_for_okta_to_aws_role_mapping(
...
+    for schema in schemas:
+        load_matchlinks(
+            neo4j_session,
+            schema,
+            mappings,
+            lastupdated=okta_update_tag,
+            _sub_resource_label="OktaOrganization",
+            _sub_resource_id=okta_org_id,
+        )
+        GraphJob.from_matchlink(
```
Comment (jychp) [inline cartography/intel/okta/awssaml.py:152]: Move the `GraphJob.from_matchlink(...).run()` calls out of `_load_okta_group_to_aws_roles` into a `cleanup()` function called from `sync_okta_aws_saml`. A loader must not delete.

File: cartography/intel/okta/awssaml.py [hunk 4]
```diff
@@ -97,32 +108,41 @@ def query_for_okta_to_aws_role_mapping(
...
-    ON CREATE SET r.firstseen = timestamp()
-    SET r.lastupdated = $okta_update_tag
-    """
-
-    run_write_query(
-        neo4j_session,
-        ingest_statement,
-        GROUP_TO_ROLE=group_to_role,
-        okta_update_tag=okta_update_tag,
+    mappings = [mapping._asdict() for mapping in group_to_role]
```
Comment (jychp) [inline cartography/intel/okta/awssaml.py]: Dedupe `group_to_role` before building `mappings`. A group assigned to both the `amazon_aws` and `amazon_aws_sso` applications resolves to the same `(okta_group_id, aws_role_arn)` pair twice and is merged twice per schema.

File: cartography/intel/okta/awssaml.py [hunk 5]
```diff
@@ -259,17 +252,26 @@ def sync_okta_aws_saml(
     group_to_role_mapping = query_for_okta_to_aws_role_mapping(
         neo4j_session,
         mapping_regex,
+        okta_org_id,
     )
-    _load_okta_group_to_aws_roles(neo4j_session, group_to_role_mapping, okta_update_tag)
+    combined_group_to_role_mapping = [
```
Comment (jychp) [inline cartography/intel/okta/awssaml.py]: Return `GroupRole` from `query_for_okta_to_aws_role_mapping` and delete this remapping block. Two shapes for the same concept (`{"groupid", "role"}` vs `GroupRole`) only exist because the loader signature changed.

File: tests/integration/cartography/intel/okta/test_awssaml.py [hunk 6]
```diff
@@ -202,6 +234,175 @@ def test_sync_okta_aws_saml_no_matching_roles(neo4j_session):
...
+    other_org_id = "other-okta-org-id"
+    other_group_id = "other-org-group"
+    other_sso_group_id = "other-org-sso-group"
+    other_role_arn = "arn:aws:iam::4321:role/other-org-role"
+    other_sso_role_arn = "arn:aws:iam::4321:role/AWSReservedSSO_other-sso-role_abcdef"
+    neo4j_session.run(
+        """
+        MERGE (current_org:OktaOrganization {id: $CURRENT_ORG_ID})
+        MERGE (org:OktaOrganization {id: $ORG_ID})
+        MERGE (app:OktaApplication {name: "amazon_aws"})
```
Comment (jychp) [inline tests/integration/cartography/intel/okta/test_awssaml.py]: This `MERGE` keys on `name` only, so both orgs share a single `OktaApplication` node and the new `(org)-[:RESOURCE]->(app)` clause is never exercised. Give each org its own app node with a distinct `id`.

File: tests/integration/cartography/intel/okta/test_awssaml.py [hunk 7]
```diff
@@ -202,6 +234,175 @@ def test_sync_okta_aws_saml_no_matching_roles(neo4j_session):
...
+    relationship_count = neo4j_session.run(
+        """
+        MATCH (group:OktaGroup)-[r]->(:AWSRole)
+        WHERE group.id IN $GROUP_IDS
+              AND type(r) IN ["ALLOWED_BY", "HAS_ROLE"]
+        RETURN count(r) AS count
+        """,
+        GROUP_IDS=[other_group_id, other_sso_group_id],
+    ).single(strict=True)["count"]
+    assert relationship_count == 0
```
Comment (jychp) [inline tests/integration/cartography/intel/okta/test_awssaml.py]: Set up current-org groups and roles too, and assert their `HAS_ROLE`/`ALLOWED_BY` edges are created. As written this test passes even if `sync_okta_aws_saml` is a no-op.

File: tests/integration/cartography/intel/okta/test_awssaml.py [hunk 8]
```diff
@@ -202,6 +234,175 @@ def test_sync_okta_aws_saml_no_matching_roles(neo4j_session):
...
+
+    # Assert
+    relationship_count = neo4j_session.run(
+        """
+        MATCH (:OktaGroup {id: $GROUP_ID})-[r:ALLOWED_BY]->(:AWSRole)
+        RETURN count(r) AS count
+        """,
+        GROUP_ID=group_id,
+    ).single(strict=True)["count"]
+    assert relationship_count == 0
```
Comment (jychp) [inline tests/integration/cartography/intel/okta/test_awssaml.py:445]: Cover the upgrade path that actually matters: a pre-existing unscoped `ALLOWED_BY` for a mapping that is still valid must be adopted by the sync (gaining `_sub_resource_label`) and survive this migration statement. Only the orphaned-edge case is tested.

**Review by jychp — CHANGES_REQUESTED** [review]

File: cartography/data/jobs/cleanup/okta_import_cleanup.json (hunk 1 above)
Comment (kunaals) [reply cartography/data/jobs/cleanup/okta_import_cleanup.json]: yep, this needs to be global. updated it to catch orphaned groups too.

File: cartography/intel/okta/awssaml.py (hunk 3 above)
Comment (kunaals) [reply cartography/intel/okta/awssaml.py:152]: moved this into a separate cleanup step after loading.

File: cartography/intel/okta/awssaml.py (hunk 4 above)
Comment (kunaals) [reply cartography/intel/okta/awssaml.py]: deduped these before loading either relationship.

File: cartography/intel/okta/awssaml.py (hunk 5 above)
Comment (kunaals) [reply cartography/intel/okta/awssaml.py]: cleaned this up so the direct saml path uses `GroupRole` throughout.

File: tests/integration/cartography/intel/okta/test_awssaml.py (hunk 6 above)
Comment (kunaals) [reply tests/integration/cartography/intel/okta/test_awssaml.py]: fixed the fixture so each org has distinct application nodes.

File: tests/integration/cartography/intel/okta/test_awssaml.py (hunk 7 above)
Comment (kunaals) [reply tests/integration/cartography/intel/okta/test_awssaml.py]: added positive assertions for both current-org paths.

File: tests/integration/cartography/intel/okta/test_awssaml.py (hunk 8 above)
Comment (kunaals) [reply tests/integration/cartography/intel/okta/test_awssaml.py:445]: added this upgrade case and confirmed the adopted edge survives cleanup.

File: cartography/data/jobs/cleanup/okta_import_cleanup.json (hunk 1 above)
Comment (kunaals) [reply cartography/data/jobs/cleanup/okta_import_cleanup.json]: i accidentally reintroduced the org scope while addressing a later review. reverted that- the migration stays global so it can clean up orphaned legacy edges, while the recurring matchlink cleanup remains scoped per okta org.

File: tests/integration/cartography/intel/okta/test_awssaml.py [hunk 9]
```diff
@@ -202,6 +234,302 @@ def test_sync_okta_aws_saml_no_matching_roles(neo4j_session):
...
+    role_arn = "arn:aws:iam::1234:role/myrole1"
+    _setup_okta_test_data(neo4j_session)
+    _setup_aws_test_data(neo4j_session)
+    neo4j_session.run(
+        """
+        MATCH (org:OktaOrganization {id: $ORG_ID})
+        MATCH (app:OktaApplication {name: "amazon_aws"})
+        MATCH (role:AWSRole {arn: $ROLE_ARN})
+        MERGE (group:OktaGroup {id: $GROUP_ID})
+        SET group.name = "aws#test#myrole1#1234"
```
[the remark this answers is not in the document]
Comment (kunaals) [reply tests/integration/cartography/intel/okta/test_awssaml.py]: fixed. the fixture now sets the current update tag, matching the real group sync before migration cleanup.

File: cartography/models/okta/awssaml.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
+from dataclasses import dataclass
```
[the remark this answers is not in the document]
Comment (kunaals) [reply cartography/models/okta/awssaml.py:1]: keeping this visible because it is a real compatibility edge that users can query until v1, matching how `MEMBER_OF_OKTA_GROUP` is documented. introspection exclusion is for schemas that don’t accurately describe a concrete edge.

File: tests/integration/cartography/intel/okta/test_awssaml.py [hunk 11]
```diff
@@ -202,6 +234,283 @@ def test_sync_okta_aws_saml_no_matching_roles(neo4j_session):
...
+
+    # Assert
+    relationship_count = neo4j_session.run(
+        """
+        MATCH (:OktaGroup {id: $GROUP_ID})-[r:ALLOWED_BY]->(:AWSRole)
+        RETURN count(r) AS count
+        """,
+        GROUP_ID=group_id,
+    ).single(strict=True)["count"]
+    assert relationship_count == 0
```
[the remark this answers is not in the document]
Comment (kunaals) [reply tests/integration/cartography/intel/okta/test_awssaml.py:445]: keeping the smaller orphaned-edge case. because the group has no `OktaOrganization` `RESOURCE` edge, this test already fails if the migration is scoped to `$OKTA_ORG_ID`, so it directly pins the global behavior. a second org would exercise the same predicate without adding coverage.
